### PR TITLE
[SPARK-38185][SQL] Fix data incorrect if aggregate function is empty

### PR DESCRIPTION
**Pull request by ulysses-you**
### What changes were proposed in this pull request?

Add `aggregateExpressions.nonEmpty` check in `groupOnly` function.

### Why are the changes needed?

The group only condition should check if the aggregate expression is empty.

In DataFrame api, it is allowed to make a empty aggregations.

So the following query should return 1 rather than 0 because it's a global aggregate.
```scala
val emptyAgg = Map.empty[String, String]
spark.range(2).where("id > 2").agg(emptyAgg).limit(1).count
```

### Does this PR introduce _any_ user-facing change?

yes, bug fix

### How was this patch tested?

Add test

### CODE REVIEW

**Comment by ulysses-you** [comment]
cc @wangyum @cloud-fan

**Review by cloud-fan — APPROVED** [review]
good catch!

**Comment by cloud-fan** [comment]
cc @sigmod as well

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala [hunk 1]
```diff
@@ -984,7 +984,8 @@ case class Aggregate(
 
   // Whether this Aggregate operator is group only. For example: SELECT a, a FROM t GROUP BY a
   private[sql] def groupOnly: Boolean = {
-    aggregateExpressions.map {
+    // aggregateExpressions can be empty through Dateset.agg
+    aggregateExpressions.nonEmpty && aggregateExpressions.map {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala:988]: is it more explicit to check `groupingExpressions.nonEmpty`? Logically, an aggregate operator without output columns is still group only if it has group columns.

Comment (ulysses-you) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala]: make sense

**Review by sigmod — APPROVED** [review]

**Review by HyukjinKwon — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master/3.2!

**Review by dongjoon-hyun — COMMENTED** [review]
+1, LGTM. Thank you, @ulysses-you and all.

**Comment by dongjoon-hyun** [comment]
Could you make a backporting PR for `branch-3.1` please, @ulysses-you ?
```scala
scala> val emptyAgg = Map.empty[String, String]
emptyAgg: scala.collection.immutable.Map[String,String] = Map()

scala> spark.range(2).where("id > 2").agg(emptyAgg).limit(1).count
res0: Long = 1

scala> spark.version
res1: String = 3.1.2
```

**Comment by dongjoon-hyun** [comment]
The code looks good but `branch-3.2` seems not to have this issue.
The UT passes without the patch in `branch-3.2`. Do you know what difference causes the issue at master branch, @ulysses-you ?
BTW, I'm good to have this in branch-3.2 in any case because new code looks robust.

**Comment by ulysses-you** [comment]
thank you @dongjoon-hyun , The rule that change aggregate if it's group only with limit 1 to project is landed at branch-3.3 see [SPARK-36183](https://issues.apache.org/jira/browse/SPARK-36183), and it's the key to trigger this bug. But the group only condition is added since branch-3.2, see [SPARK-34808](https://issues.apache.org/jira/browse/SPARK-34808).

So it would be good to also fix this in branch-3.2.

**Pull request closed without merging**